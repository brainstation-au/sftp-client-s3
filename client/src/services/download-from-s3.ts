import { S3 } from 'aws-sdk';
import * as fs from 'fs';

export const downloadFromS3 = async (bucket: string, key: string, localPath: string): Promise<void> => {
  const s3 = new S3({apiVersion: '2006-03-01'});
  const request = s3.getObject({
    Bucket: bucket,
    Key: key,
  });
  const w = fs.createWriteStream(localPath);

  return new Promise((resolve, reject) => {
    request.createReadStream()
      .pipe(w)
      .on('close', resolve)
      .on('error', reject);
  });
};
