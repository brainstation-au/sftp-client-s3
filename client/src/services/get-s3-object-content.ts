import { S3 } from 'aws-sdk';
import { parseS3Uri } from './parse-s3-uri';

export const getS3ObjectContent = async (uri: string): Promise<string> => {
  const {bucket, key} = parseS3Uri(uri);
  console.log(bucket,key);

  const s3 = new S3({apiVersion: '2006-03-01'});
  const data = await s3.getObject({
    Bucket: bucket,
    Key: key,
  }).promise();

  return (data.Body as Buffer).toString('utf-8');
};
