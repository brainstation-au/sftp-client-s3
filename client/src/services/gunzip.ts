import * as fs from 'fs';
import * as zlib from 'zlib';

export const gunzip = async (compressedFilepath: string, uncompressedFilepath: string): Promise<void> => {
  const r = fs.createReadStream(compressedFilepath);
  const z = zlib.createGunzip();
  const w = fs.createWriteStream(uncompressedFilepath);
  return new Promise((resolve, reject) => {
    r.pipe(z).pipe(w).on('close', resolve).on('error', reject);
  });
};
