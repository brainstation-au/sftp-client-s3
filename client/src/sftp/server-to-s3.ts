import { S3 } from 'aws-sdk';
import * as fs from 'fs';
import * as path from 'path';
import Client from 'ssh2-sftp-client';
import { ServerParams } from '../types/server-params';
import { ServerToS3Options } from '../types/server-to-s3-options';
import { execute } from './sftp';

export const serverToS3 = async (options: ServerToS3Options): Promise<void> => {
  console.log(options);
};

export const downloadFromSftpServer = async (options: ServerParams): Promise<string[]> => {
  const {
    host,
    port,
    username,
    privateKey,
    location,
    filename,
  } = options;
  const storageLocation = process.env['STORAGE_LOCATION'] || '/tmp/download/';
  fs.mkdirSync(storageLocation, {recursive: true});
  const downloadLocation = fs.mkdtempSync(storageLocation);

  return execute({host, port, username, privateKey}, async (client: Client): Promise<string[]> => {
    const files: Array<{remotePath: string; localPath: string}> = await (filename ? client.list(location, filename) : client.list(location))
      .then(l => l.map(f => ({
        remotePath: path.join(location, f.name),
        localPath: path.join(downloadLocation, f.name),
      })));
    for (const file of files) {
      await client.fastGet(file.remotePath, file.localPath);
    }

    return files.map(f => f.localPath);
  });
};

export const uploadToS3 = async (localPath: string, bucket: string, key: string): Promise<S3.Types.PutObjectOutput> => {
  const s3 = new S3({apiVersion: '2006-03-01'});
  return s3.putObject({
    Body: fs.readFileSync(localPath),
    Bucket: bucket,
    Key: key,
  }).promise();
};
