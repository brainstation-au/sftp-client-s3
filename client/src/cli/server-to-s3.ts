import { Arguments, Argv } from 'yargs';
import { serverToS3 } from '../sftp/server-to-s3';
import { ServerToS3Options } from '../types/server-to-s3-options';

export const command = 'server-to-s3';

export const description = 'Download files from SFTP server and put them in S3 bucket';

export const builder = (yargs: Argv<unknown>): Argv<unknown> => yargs
  .option('host', {
    alias: ['sftp-host', 'h'],
    default: process.env['SFTP_HOST'],
    demandOption: !('SFTP_HOST' in process.env),
    description: 'SFTP host IP address or URL',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('port', {
    alias: ['sftp-port', 'p'],
    default: process.env['SFTP_PORT'],
    demandOption: !('SFTP_PORT' in process.env),
    description: 'SFTP host port number',
    nargs: 1,
    requiresArg: true,
    type: 'number',
  })
  .option('user', {
    alias: ['sftp-user', 'username', 'u'],
    default: process.env['SFTP_USER'],
    demandOption: !('SFTP_USER' in process.env),
    description: 'SFTP username',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('key', {
    alias: ['private-key', 'k'],
    default: process.env['PRIVATE_KEY'],
    demandOption: !('PRIVATE_KEY' in process.env),
    description: 'Private key to authenticate SFTP session',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('location', {
    alias: ['remote-location', 'l'],
    default: process.env['REMOTE_LOCATION'],
    demandOption: !('REMOTE_LOCATION' in process.env),
    description: 'Path to the file location in SFTP server',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('filename', {
    alias: ['filename-pattern', 'f'],
    default: process.env['FILENAME'],
    description: 'Name of the file or a regular expression to find a subset',
    nargs: 1,
    type: 'string',
  })
  .option('bucket', {
    alias: ['bucket-name', 'b'],
    default: process.env['BUCKET_NAME'],
    demandOption: !('BUCKET_NAME' in process.env),
    description: 'S3 bucket name',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('key-prefix', {
    alias: ['s3-key-prefix', 's3-key'],
    default: process.env['KEY_PREFIX'],
    demandOption: !('KEY_PREFIX' in process.env),
    description: 'S3 key prefix to upload the file',
    nargs: 1,
    requiresArg: true,
    type: 'string',
  })
  .option('decrypt', {
    alias: ['d'],
    default: process.env['DECRYPT'] || false,
    description: 'Decrypt file content with PGP private key',
    type: 'boolean',
  });

export const handler = async (argv: Arguments): Promise<void> => {
  const options = ServerToS3Options.check(argv);
  return serverToS3(options);
};
