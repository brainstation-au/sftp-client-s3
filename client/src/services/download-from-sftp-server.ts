import Client from 'ssh2-sftp-client';
import { ServerParams } from '../types/server-params';
import { execute } from './sftp';

export const downloadFromSftpServer = async (options: ServerParams, remotePath: string, localPath: string): Promise<string> => {
  return execute(options, async (client: Client): Promise<string> => {
    return client.fastGet(remotePath, localPath);
  });
};
