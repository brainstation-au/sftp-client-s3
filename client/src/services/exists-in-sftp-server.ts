import Client from 'ssh2-sftp-client';
import { ServerParams } from '../types/server-params';
import { execute } from './sftp';

export const existsInSftpServer = async (options: ServerParams, remotePath: string): Promise<boolean> => {
  const {
    host,
    port,
    username,
    privateKey,
  } = options;

  return execute({host, port, username, privateKey}, async (client: Client): Promise<boolean> => {
    return client.exists(remotePath).then(res => !!res);
  });
};
