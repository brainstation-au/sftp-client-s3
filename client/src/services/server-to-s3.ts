import * as fs from 'fs';
import moment from 'moment-timezone';
import * as path from 'path';
import * as zlib from 'zlib';
import { ServerToS3Options } from '../types/server-to-s3-options';
import { downloadFromSftpServer } from './download-from-sftp-server';
import { listFromSftpServer } from './list-from-sftp-server';
import { uncompress } from './gzip';
import { localStorageLocation } from './local-storage-location';
import { decrypt } from './openpgp';
import { removeFromSftpServer } from './remove-from-sftp-server';
import { uploadToS3 } from './upload-to-s3';

export const serverToS3 = async (options: ServerToS3Options): Promise<void> => {
  const filenames = await listFromSftpServer(options);
  const localDir = localStorageLocation();
  const keyPrefix = moment().tz(options.timezone).format(options.keyPrefixPattern);

  for (const filename of filenames) {
    const isGzipped = path.extname(filename).toLowerCase() === '.gz';
    const remotePath = path.join(options.location, filename);
    const filepath = path.join(localDir, filename);
    const downloadResponse = await downloadFromSftpServer(options, remotePath, filepath);
    console.log(downloadResponse);

    if (options.decrypt) {
      const fileContent = fs.readFileSync(filepath);
      const encryptedContent: Buffer = isGzipped ? zlib.gunzipSync(fileContent) : fileContent;
      const clearContent: string = await decrypt(encryptedContent.toString(), options.gpgPrivateKey, options.gpgPassphrase);
      const contentToWrite: Buffer = isGzipped ? zlib.gzipSync(Buffer.from(clearContent)) : Buffer.from(clearContent);
      fs.writeFileSync(filepath, contentToWrite);
    }

    let uploadName = filename;
    if (isGzipped && options.gunzip) {
      uploadName = path.parse(filename).name;
      await uncompress(path.join(localDir, filename), path.join(localDir, uploadName));
      fs.unlinkSync(path.join(localDir, filename));
    }

    await uploadToS3(path.join(localDir, uploadName), options.bucket, `${keyPrefix}${uploadName}`);
    console.log(`s3://${options.bucket}/${keyPrefix}${uploadName} has been uploaded.`);
    fs.unlinkSync(path.join(localDir, uploadName));

    if (options.rm) {
      await removeFromSftpServer(options, filename);
      console.log(`${filename} has been deleted from the server.`);
    }
  }
};
