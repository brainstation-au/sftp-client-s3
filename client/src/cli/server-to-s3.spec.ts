import yargs, { Arguments } from 'yargs';
import * as serverToS3 from './server-to-s3';
import * as handler from '../services/server-to-s3';
jest.mock('../services/server-to-s3');
const mockedHandler = handler as jest.Mocked<typeof handler>;

describe('server-to-s3', () => {
  const parser = yargs.command(serverToS3).help();
  const parseArgs = (args: string): Promise<string> => new Promise((resolve, reject) => {
    parser.parse(args, (err: Error | undefined, _argv: Arguments, output: string): void => {
      if (err) reject(err);
      resolve(output);
    });
  });

  describe('help', () => {
    test('basic help output on parent', async () => {
      await expect(parseArgs('--help')).resolves.toMatchSnapshot();
    });

    test('basic help output on command', async () => {
      await expect(parseArgs('server-to-s3 --help')).resolves.toMatchSnapshot();
    });
  });

  describe('command options', () => {
    describe('required params are not available', () => {
      test('with no arguments', async () => {
        await expect(parseArgs('server-to-s3')).rejects
          .toThrow('Missing required arguments: host, user, key, location, bucket, key-prefix');
      });

      test('with few arguments', async () => {
        await expect(parseArgs('server-to-s3 --host sftphost --location /download/from/here -b test-bucket')).rejects
          .toThrow('Missing required arguments: user, key, key-prefix');
      });
    });

    describe('required params are available', () => {
      describe('params are available through environment', () => {
        let output: string;

        beforeAll(async () => {
          mockedHandler.serverToS3.mockResolvedValue();
          process.env['SFTP_HOST'] = 'sftphost';
          process.env['SFTP_USER'] = 'test_user';
          process.env['KEY_PREFIX'] = 'upload/here';

          output = await parseArgs('server-to-s3 --port 22 --k secret-code --location /download/from/here -b test-bucket');
        });

        afterAll(() => {
          mockedHandler.serverToS3.mockClear();
          delete process.env['SFTP_HOST'];
          delete process.env['SFTP_USER'];
          delete process.env['KEY_PREFIX'];
        });

        test('resolves successfully', () => {
          expect(output).toEqual('');
        });

        test('calls handler with arguments', async () => {
          expect(mockedHandler.serverToS3).toHaveBeenCalledTimes(1);
          expect(mockedHandler.serverToS3).toHaveBeenCalledWith(expect.objectContaining({
            host: 'sftphost',
            port: 22,
            user: 'test_user',
            key: 'secret-code',
            location: '/download/from/here',
            filename: undefined,
            bucket: 'test-bucket',
            keyPrefix: 'upload/here',
            decrypt: false,
          }));
        });
      });
    });
  });
});
