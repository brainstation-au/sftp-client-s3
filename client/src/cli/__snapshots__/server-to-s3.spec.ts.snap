// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-to-s3 help basic help output on command 1`] = `
"jest server-to-s3

Download files from SFTP server and put them in S3 bucket

Options:
      --version                        Show version number             [boolean]
      --help                           Show help                       [boolean]
  -h, --host, --sftp-host              SFTP host IP address or URL
                                                             [string] [required]
  -p, --port, --sftp-port              SFTP host port number
                                                          [number] [default: 22]
  -u, --user, --sftp-user, --username  SFTP username         [string] [required]
  -k, --key, --private-key             Private key to authenticate SFTP session
                                                             [string] [required]
  -l, --location, --remote-location    Path to the file location in SFTP server
                                                             [string] [required]
  -f, --filename, --filename-pattern   Name of the file or a regular expression
                                       to find a subset                 [string]
  -b, --bucket, --bucket-name          S3 bucket name        [string] [required]
      --key-prefix, --s3-key-prefix    S3 key prefix to upload the file
                                                             [string] [required]
  -d, --decrypt                        Decrypt file content with PGP private key
                                                      [boolean] [default: false]"
`;

exports[`server-to-s3 help basic help output on parent 1`] = `
"jest [command]

Commands:
  jest server-to-s3  Download files from SFTP server and put them in S3 bucket

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]"
`;
